job-runner/chain/chain.go

```diff
@@ -292,66 +371,136 @@ func (c *chain) SetJobState(jobId string, state byte) {
+		// run stays correct.
+		if c.numJobsRun == 0 {
+			// Don't decrement below 0 - n is unsigned
+			return
+		}
+		c.numJobsRun--
+	}
 }
 
 // Running returns a list of running jobs.
-func (c *chain) Running() map[string]proto.JobStatus {
+func (c *Chain) Running() map[string]proto.JobStatus {
 	// Return copy of c.running
-	c.runningMux.Lock()
-	defer c.runningMux.Unlock()
+	c.runningMux.RLock()
+	defer c.runningMux.RUnlock()
 	running := make(map[string]proto.JobStatus, len(c.running))
 	for jobId, jobStatus := range c.running {
 		running[jobId] = jobStatus
 	}
 	return running
 }
 
+// Length returns the total number of jobs in the chain.
+func (c *Chain) Length() int {
+	return len(c.jobChain.Jobs)
+}
+
 // //////////////////////////////////////////////////////////////////////////
-// Implement JOSN interfaces for custom (un)marshalling by chain.Repo
+// Implement JSON interfaces for custom (un)marshalling by chain.Repo
 // //////////////////////////////////////////////////////////////////////////
 
 type chainJSON struct {
-	JobChain *proto.JobChain            `json:"jobChain"`
-	Running  map[string]proto.JobStatus `json:"running"`
-	N        uint                       `json:"n"`
+	// An SJC has already fields for most of the info we need to store.
+	proto.SuspendedJobChain
+	Running    map[string]proto.JobStatus `json:"running"`
+	NumJobsRun uint                       `json:"numJobsRun"`
 }
 
-func (c *chain) MarshalJSON() ([]byte, error) {
-	c.jcMux.Lock()
-	defer c.jcMux.Unlock()
-
-	c.runningMux.Lock()
-	defer c.runningMux.Unlock()
+func (c *Chain) MarshalJSON() ([]byte, error) {
+	c.runningMux.RLock()
+	running := c.running
+	numJobsRun := c.numJobsRun
+	c.runningMux.RUnlock()
 
 	m := chainJSON{
-		JobChain: c.jobChain,
-		Running:  c.running,
-		N:        c.n,
+		SuspendedJobChain: c.ToSuspended(),
+		Running:           running,
+		NumJobsRun:        numJobsRun,
 	}
 	return json.Marshal(m)
 }
 
-func (c *chain) UnmarshalJSON(bytes []byte) error {
+func (c *Chain) UnmarshalJSON(bytes []byte) error {
 	var m chainJSON
 	err := json.Unmarshal(bytes, &m)
 	if err != nil {
 		return err
 	}
 
-	c.jcMux = &sync.RWMutex{}
+	c.jobsMux = &sync.RWMutex{}
 	c.jobChain = m.JobChain
-
+	c.triesMux = &sync.RWMutex{}
+	c.sequenceTries = m.SequenceTries
+	c.totalJobTries = m.TotalJobTries
+	c.latestRunJobTries = m.LatestRunJobTries
+	c.numJobsRun = m.NumJobsRun
 	c.runningMux = &sync.RWMutex{}
 	c.running = m.Running
-	c.n = m.N
 
 	return nil
 }
 
 // -------------------------------------------------------------------------- //
 
+// isRunnable returns whether or not a job is runnable. A job is considered
+// runnable if it is Pending or Stopped with some retry attempts remaining,
+// and all of its previous jobs are complete. If any previous jobs are not
+// complete, the job is not unnable.
+// Doesn't lock jobsMux, so suitable to call within other Chain methods lock jobsMux.
```

Comment:
*other Chain methods that lock

Is it only safe to call this method if you've already locked jobsMux? If so, mention that